data-products/src/main/scala/org/sunbird/analytics/util/TextBookUtils.scala

```diff
@@ -0,0 +1,183 @@
+        val data = response.result.content
+
+        val etbTextBookResponse = generateETBTextbookReport(data)
+        etbTextBookReport = etbTextBookReport ++ etbTextBookResponse
+
+        val dceTextBookResponse = generateDCETextbookReport(data)
+        dceTextBookReport = dceTextBookReport ++ dceTextBookResponse
+      }
+    })
+    val tenantRDD = tenantInfo.map(e=>(e.id,e))
+
+    val etbTextBook = sc.parallelize(etbTextBookReport).map(e=>(e.channel,e))
+    val etb=etbTextbookData("","","","","","","","","",0,0,0,0,0)
+    val etbTextBookRDD = etbTextBook.fullOuterJoin(tenantRDD).map(etbReport => {
+
+      etbTextbookReport(etbReport._2._2.getOrElse(TenantInfo("","unknown")).slug, etbReport._2._1.getOrElse(etb).identifier,
+        etbReport._2._1.getOrElse(etb).name,etbReport._2._1.getOrElse(etb).medium,etbReport._2._1.getOrElse(etb).gradeLevel,
+        etbReport._2._1.getOrElse(etb).subject,etbReport._2._1.getOrElse(etb).status,etbReport._2._1.getOrElse(etb).createdOn,etbReport._2._1.getOrElse(etb).lastUpdatedOn,
+        etbReport._2._1.getOrElse(etb).totalContentLinked,etbReport._2._1.getOrElse(etb).totalQRLinked,etbReport._2._1.getOrElse(etb).totalQRNotLinked,
+        etbReport._2._1.getOrElse(etb).leafNodesCount,etbReport._2._1.getOrElse(etb).leafNodeUnlinked,"ETB_textbook_data")
+    })
+
+    val dceTextBook = sc.parallelize(dceTextBookReport.filter(e=>(e.totalQRCodes!=0))).map(e=>(e.channel,e))
+    val dce=dceTextbookData("","","","","","","","",0,0,0,0,0)
+    val dceTextBookRDD = dceTextBook.fullOuterJoin(tenantRDD).map(dceReport => {
+      dceTextbookReport(dceReport._2._2.getOrElse(TenantInfo("","unknown")).slug,dceReport._2._1.getOrElse(dce).identifier,
+        dceReport._2._1.getOrElse(dce).name,dceReport._2._1.getOrElse(dce).medium,dceReport._2._1.getOrElse(dce).gradeLevel,dceReport._2._1.getOrElse(dce).subject,
+        dceReport._2._1.getOrElse(dce).createdOn,dceReport._2._1.getOrElse(dce).lastUpdatedOn,dceReport._2._1.getOrElse(dce).totalQRCodes,
+        dceReport._2._1.getOrElse(dce).contentLinkedQR,dceReport._2._1.getOrElse(dce).withoutContentQR,dceReport._2._1.getOrElse(dce).withoutContentT1,
+        dceReport._2._1.getOrElse(dce).withoutContentT2,"DCE_textbook_data")
+    })
+
+    val dceRDD=dceTextBookRDD.map(e=>(e.identifier,e))
+    val etbRDD=etbTextBookRDD.map(e=>(e.identifier,e)).fullOuterJoin(dceRDD)
+    val finalOutput = etbRDD.map(e=> FinalOutput(e._1,e._2._1,e._2._2))
+
+    finalOutput
+  }
+
+  def generateDCETextbookReport(response: ContentInfo)(implicit sc: SparkContext): List[dceTextbookData] = {
+    var index=0
+    var totalQRCodes=0
+    var qrLinked=0
+    var qrNotLinked=0
+    var term1NotLinked=0
+    var term2NotLinked=0
+    var dceReport = List[dceTextbookData]()
+
+    if(response!=null && response.children.size>0 && response.status=="Live") {
+      val lengthOfChapters = response.children.get.length
+      val term = if(index<=lengthOfChapters/2) "T1"  else "T2"
```

Comment:
Shouldn't this be within the chapter loop?

Reply:
If it remains within the chapter loop, I will not be able to calculate the term for each chapter.